044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,198 @@
+
+The broker nodes cannot catch up with the cluster metadata until the controller quorum has been formed, so they will not move out of the STARTING state until the quorum leader election has happened.
+
+## Proposal
+
+There are three different types of liveness/readiness probe that we need to consider, first for a controller only, then 
+for a broker only, and finally for a combined node that has both roles.
+
+The following statements describe the intent of the proposed probes:
+
+* A controller is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to join the controller quorum.
+* A broker is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+* A combined node is:
+    * "alive" if it has a process running.
+    * "ready" when it is ready to start accepting producer/consumer requests.
+        * This implies that the controller quorum will have been formed while the nodes are not marked as ready, since it cannot start accepting producer/consumer requests until it can talk to the leader.
+        * This relies on the node still accepting incoming connections from other controller nodes even if it isn't actually marked as "ready" yet.
+
+### Controller only mode
+
+The proposed probes are:
+
+* Liveness: the Java process in the container is running
+* Readiness: controller is listening on the port 9090 which is configured in `controller.listener.names`
+
+### Broker only mode
+
+The proposed probes are:
+
+* Liveness: broker is listening on the port 9091 which is configured in `inter.broker.listener.name`
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the brokers will not become ready until a majority of the controllers 
+are up and running.
+This is similar to the current behaviour of brokers when ZooKeeper is not ready.
+
+### Combined mode
+
+The proposed probes are:
+
+* Liveness: the Java process in the container is running
+* Readiness: the BrokerState metric >=3 && != 127
+
+**Note:** This means the nodes will not become ready until a majority of the other controllers are up and running.
+This is acceptable because the Strimzi headless services use "publishNotReadyAddresses", which 
+means the nodes will be able to communicate even if they are currently marked as not ready.
+
+### Implementation of the checks
+
+To check that the Java process is running we will execute a command in the container. The proposed command is:
+```shell
+for proc in $(ls -1 /proc/ | grep [0-9]); do echo "$(ls -lh /proc/$proc/exe 2>/dev/null || true)" | grep -q java; done
+```
+This is preferable over using a tool like `ps` (`ps aux | grep -v grep | grep java`) since it does not require a change to the tools added to the container image.
+In future this could be replaced with a call to an endpoint, since other proposals discuss adding endpoints for example to expose metrics.
```

Comment:
```suggestion
In future this could be replaced with a call to an endpoint, since other proposals discuss adding endpoints, for example, to expose metrics.
```